File: src/runtime/context.ts
import { ExecutionTrace } from "./trace.js";
import { VariableStore } from "./variables.js";

export interface RuntimeContextOptions {
  context?: Record<string, unknown>;
  mockFS?: Record<string, unknown>;
}

export class RuntimeContext {
  readonly context: Record<string, unknown>;
  readonly variables = new VariableStore();
  readonly trace = new ExecutionTrace();
  readonly outputs: unknown[] = [];
  private readonly mockFS = new Map<string, unknown>();

  constructor(options: RuntimeContextOptions = {}) {
    this.context = options.context ?? {};

    for (const [path, value] of Object.entries(options.mockFS ?? {})) {
      this.mockFS.set(path, value);
    }
  }

  readFile(path: string): unknown {
    return this.mockFS.get(path);
  }

  writeFile(path: string, content: unknown): void {
    this.mockFS.set(path, content);
  }

  snapshotFS(): Record<string, unknown> {
    return Object.fromEntries(this.mockFS.entries());
  }
}

File: src/runtime/context.rs
use std::collections::BTreeMap;

use serde_json::{Map, Value};

use super::trace::ExecutionTrace;
use super::variables::VariableStore;

#[derive(Debug, Default)]
pub struct RuntimeContext {
    pub context: Map<String, Value>,
    pub variables: VariableStore,
    pub trace: ExecutionTrace,
    pub outputs: Vec<Value>,
    mock_fs: BTreeMap<String, Value>,
}

impl RuntimeContext {
    pub fn new(context: Map<String, Value>, mock_fs: BTreeMap<String, Value>) -> Self {
        Self {
            context,
            variables: VariableStore::default(),
            trace: ExecutionTrace::default(),
            outputs: Vec::new(),
            mock_fs,
        }
    }

    pub fn read_file(&self, path: &str) -> Option<&Value> {
        self.mock_fs.get(path)
    }

    pub fn write_file(&mut self, path: impl Into<String>, content: Value) {
        self.mock_fs.insert(path.into(), content);
    }

    pub fn snapshot_fs(&self) -> BTreeMap<String, Value> {
        self.mock_fs.clone()
    }
}
